Wenzhe/llga fuser

wenzhe-nrv opened the pull request:
for internal review.

wenzhe-nrv committed:
add fuser3 context manager for onednn graph

wenzhe-nrv committed:
add oneDNN graph fuser context test

wenzhe-nrv committed:
minor edit

wenzhe-nrv committed:
Merge branch 'pytorch:master' into wenzhe/llga_fuser

chunyuan-w commented on torch/jit/_fuser.py:
```diff
@@ -48,6 +50,12 @@ def fuser(name):
         torch._C._jit_set_texpr_fuser_enabled(False)
         torch._C._jit_set_nvfuser_enabled(True)
+    elif name == 'fuser3': # oneDNN Graph
+        torch._C._jit_override_can_fuse_on_cpu(old_cpu_fuse)
```
May I know why we need to set the state to `old_cpu_fuse` here?

wenzhe-nrv replied:
I was not sure about this and chose to keep the previous setting earlier. Same for your comment below. 
Are you suggesting to set both `True` like nnc?

chunyuan-w replied:
Oh I was just asking since I haven't seen such settings in 'fuser0' `fuser1` or `fuser2` to keep the old status.
I think it should be fine to just set the old status back in the `finally` statement.

chunyuan-w commented on torch/jit/_fuser.py:
```diff
@@ -48,6 +50,12 @@ def fuser(name):
+    elif name == 'fuser3': # oneDNN Graph
+        torch._C._jit_override_can_fuse_on_cpu(old_cpu_fuse)
+        torch._C._jit_override_can_fuse_on_gpu(False)
+        torch._C._jit_set_texpr_fuser_enabled(old_texpr_fuser_state)
```
May I know why we need to set the state to `old_texpr_fuser_state` here?

chunyuan-w commented on torch/jit/_fuser.py:
```diff
@@ -48,6 +50,12 @@ def fuser(name):
         torch._C._jit_set_texpr_fuser_enabled(False)
         torch._C._jit_set_nvfuser_enabled(True)
+    elif name == 'fuser3': # oneDNN Graph
+        torch._C._jit_override_can_fuse_on_cpu(old_cpu_fuse)
```
Similar to NNC, oneDNN graph fuser also needs to have the profiling executor turned on. You can consider adding the below lines here as well (similar to `'fuser1'`) to guarantee that oneDNN graph fuser functions as expected:
```python
old_profiling_executor = torch._C._jit_set_profiling_executor(True)
old_profiling_mode = torch._C._get_graph_executor_optimize(True)
```

(If the profiling executor is turned on here, you'll need to recover the profiling executor state to the old one in the `finally` statement as well for oneDNN graph fuser.)

chunyuan-w commented on test/test_jit_llga_fuser.py:
```diff
@@ -481,6 +482,49 @@ def forward(self, x, y):
+                t_jit = torch.jit.script(t1)
+                t_jit(x, y)
+                t_jit(x, y)
+                self.assertGraphContains(t_jit.graph_for(x, y), LLGA_FUSION_GUARD)
```
May I know if it is possible to check
```suggestion
                self.assertGraphContains(t_jit.graph_for(x, y), LLGA_FUSION_GROUP)
```
If it works, since `LLGA_FUSION_GROUP` is already define in this file, you don't need to add a new string `LLGA_FUSION_GUARD`.

wenzhe-nrv replied:
The test passes with changing to FUSION_GROUP. 

What's the right/preferred usage for FUSION_GROUP and FUSION_GUARD?
I referred this test for fuser2 which uses FUSION_GUARD: https://github.com/wenzhe-nrv/pytorch/blob/737dbfe3b28e799603b1b992d170df4cbfb28c20/test/test_jit_cuda_fuser.py#L4902

chunyuan-w replied:
Either of them should be fine.
Since `LLGA_FUSION_GROUP` is already defined, we might not need to introduce another string.

Taken the below graph as an example, when the oneDNN graph fuser functions as expected, both `prim::oneDNNFusionGuard` and `prim::oneDNNFusionGroup` can be found in the graph.
```
graph(%self : __torch__.torch.nn.modules.pooling.___torch_mangle_379.AvgPool2d,
      %input : Tensor):
  %2 : Tensor, %3 : bool = prim::oneDNNFusionGuard[types=[Float(1, 4, 18, 18, strides=[1296, 324, 18, 1], requires_grad=0, device=cpu)]](%input)
  %4 : Tensor = prim::If(%3)
    block0():
      %5 : Tensor = prim::oneDNNFusionGroup_0[output_layouts=[0]](%2)
      -> (%5)
    block1():
      %6 : Function = prim::Constant[name="fallback_function", fallback=1]()
      %7 : (Tensor) = prim::CallFunction(%6, %input)
      %8 : Tensor = prim::TupleUnpack(%7)
      -> (%8)
  return (%4)
with prim::oneDNNFusionGroup_0 = graph(%input : Float(1, 4, 18, 18, strides=[1296, 324, 18, 1], requires_grad=0, device=cpu)):
  %1 : int[] = prim::Constant[value=[5, 5]]()
  %2 : int[] = prim::Constant[value=[1, 1]]()
  %3 : int[] = prim::Constant[value=[0, 0]]()
  %4 : bool = prim::Constant[value=0]()
  %5 : bool = prim::Constant[value=1]()
  %6 : NoneType = prim::Constant()
  %7 : Float(1, 4, 14, 14, strides=[784, 196, 14, 1], requires_grad=0, device=cpu) = aten::avg_pool2d(%input, %1, %2, %3, %4, %5, %6) # /home/chunyuan/TE/pytorch/torch/nn/modules/pooling.py:623:0
  return (%7)
```

chunyuan-w commented on test/test_jit_llga_fuser.py:
```diff
@@ -481,6 +482,49 @@ def forward(self, x, y):
             graph, _ = self.checkTrace(m, [x, y])
 
 
+class TestEnableDisableLlgaFuser(JitTestCase):
```
I suppose you need to add this decorator to make sure that the test is only executed when LLGA is enabled:
```python
@unittest.skipIf(LLGA_NOT_ENABLED, "MKL-DNN build is disabled").
```
Similar to [test_jit_llga_fuser.py#L68-L69](https://github.com/pytorch/pytorch/blob/6e9b0dcdc4a1221e2b542d2b1af5b1755b365ff5/test/test_jit_llga_fuser.py#L68-L69).

wenzhe-nrv committed:
apply feedbacks:
- use profiling executor and mode
- set back to old executor and mode state
- set unit test for mkldnn build only

wenzhe-nrv commented:
@chunyuan-w Thank you for reviewing. 

Sounds like when enabling llga, we want to enable nnc as well. So I updated the code with explicit setting:
```
    elif name == 'fuser3': # oneDNN Graph
         old_profiling_executor = torch._C._jit_set_profiling_executor(True)
         old_profiling_mode = torch._C._get_graph_executor_optimize(True)
         torch._C._jit_override_can_fuse_on_cpu(True)
         torch._C._jit_override_can_fuse_on_gpu(False)
         torch._C._jit_set_texpr_fuser_enabled(True)
         torch._C._jit_set_nvfuser_enabled(False)
         torch._C._jit_set_llga_enabled(True)
```
Does this align with your suggestion?

chunyuan-w commented on torch/jit/_fuser.py:
```diff
@@ -26,41 +26,55 @@ def fuser(name):
         torch._C._jit_override_can_fuse_on_cpu(False)
         torch._C._jit_override_can_fuse_on_gpu(False)
         torch._C._jit_set_texpr_fuser_enabled(False)
+        torch._C._jit_set_llga_enabled(False)
```
nit: suggest putting this line after `torch._C._jit_set_nvfuser_enabled(True)` here so that the order is aligned with the other fuser.

chunyuan-w commented:
> @chunyuan-w Thank you for reviewing.
> 
> Sounds like when enabling llga, we want to enable nnc as well. So I updated the code with explicit setting:
> 
> ```
>     elif name == 'fuser3': # oneDNN Graph
>          old_profiling_executor = torch._C._jit_set_profiling_executor(True)
>          old_profiling_mode = torch._C._get_graph_executor_optimize(True)
>          torch._C._jit_override_can_fuse_on_cpu(True)
>          torch._C._jit_override_can_fuse_on_gpu(False)
>          torch._C._jit_set_texpr_fuser_enabled(True)
>          torch._C._jit_set_nvfuser_enabled(False)
>          torch._C._jit_set_llga_enabled(True)
> ```
> 
> Does this align with your suggestion?

The graph will firstly run into LLGA passes and then NNC passes. I think it is fine to turn on NNC since it won't impact the fusions of oneDNN graph.

wenzhe-nrv committed:
check LLGA_FUSION_GROUP in UT

wenzhe-nrv commented:
@chunyuan-w I have updated and applied your feedbacks. Please review again.

wenzhe-nrv committed:
remove one blank line
